content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,208 @@
+
+Building and running the `folks` attribute with `nix-build -A folks` will again produce a new version of the script.
+It will print, as you may expect, `hello folks`.
+
+All the other parameters will remain the same as they have been when `hello` was
+instantiated.
+
+This is especially useful and often seen on packages that provide many
+options to customize the build.
+
+An example to mention here is the [`neovim`](https://search.nixos.org/packages?channel=22.05&show=neovim&from=0&size=50&sort=relevance&type=packages&query=neovim) attribute in nixpkgs, which has has 
+some overrideable arguments like `extraLuaPackages`, `extraPythonPackages`, or
+`withRuby`.
+
+## 3. Benefit: modifiable
+
+And now I want to introduce one of my favorite benefits:
+
+You can actually create your own version of `callPackage`. This comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+> **Note**
+> In the next examples I will not implement or show the "called" 
+> files, as I think they are not necessary to understand the point I
+> want to make.
+
+Consider the following attribute set of derivations:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  a = pkgs.callPackage ./a.nix { };
+  b = pkgs.callPackage ./b.nix { inherit a; };
+  c = pkgs.callPackage ./c.nix { inherit b; };
+  d = pkgs.callPackage ./d.nix { };
+  e = pkgs.callPackage ./e.nix { inherit c d; };
+}
+```
+
+Note that `inherit a;` is equivalent to `a = a;`.
+That is, we're passing previously declared derivations as arguments to other derivations through `callPackage`.
+
+In this case you have to remember to manually pass arguments required by each package in the respective `.nix` file if they are not in `nixpkgs`.
+This is due to how `pkgs.callPackage` works: it passes attributes that exist in `pkgs` to the called function if the argument names match.
+
+If `./b.nix` requires an argument `a` but there is no `pkgs.a`, the function call will produce an error.
+
+This can become quite tedious quickly, especially the larger the set becomes.
+
+Therefore we can use `lib.callPackageWith` to create our own `callPackage`:
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+  callPackage = lib.callPackageWith (pkgs // packages);
+  packages = {
+    a = callPackage ./a.nix { };
+    b = callPackage ./b.nix { };
+    c = callPackage ./c.nix { };
+    d = callPackage ./d.nix { };
+    e = callPackage ./e.nix { };
+  };
+in
+  packages
+```
+
+Our modified `callPackage` now will exactly "know" how to resolve the dependencies
+through the set defined by `pkgs // packages`.
+
+Nix' laziness does us a good favour here and makes this actually possible.
+
+## Summary
+
+Using `callPackage` does not only follow `nixpkgs` conventions, making your code easy to follow for experienced Nix users. It also gives you some things for free:
```

Comment:
```suggestion
Using `callPackage` is not only a `nixpkgs` conventions.
It makes your code easy to follow for experienced Nix users, and it gives you some benefits for free:
```

Reply:
My point here was that following conventions is what makes your code readable for other Nixers. Following conventions is not a benefit on its own.

@MMesch Does that make it clearer?
```suggestion
Using `callPackage` does not only follow `nixpkgs` conventions, which makes your code easier to follow for experienced Nix users. It also gives you some benefits for free:
```

Reply:
yes, lgtm.